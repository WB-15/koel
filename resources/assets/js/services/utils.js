export default {
    /**
     * Convert a duration in seconds into H:i:s format.
     * If H is 0, it will be ommited.
     */
    secondsToHis(d) {
        d = parseInt(d);
        
        var s = d%60;

        if (s < 10) {
            s = '0' + s;
        }

        var i = Math.floor((d/60)%60);

        if (i < 10) {
            i = '0' + i;
        }

        var h = Math.floor(d/3600);

        if (h < 10) {
            h = '0' + h;
        }

        return (h === '00' ? '' : h + ':') + i + ':' + s;
    },

    /**
     * Parse the validation error from the server into a flattened array of messages.
     * 
     * @param  {Object}  error  The error object in JSON format.
     *                                    
     * @return {Array}
     */
    parseValidationError(error) {
        return Object.keys(error).reduce((messages, field) => messages.concat(error[field]), []);
    }
};
