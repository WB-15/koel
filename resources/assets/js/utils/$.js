export const $ = {
  is (el, selector) {
    return (el.matches ||
      el.matchesSelector ||
      el.msMatchesSelector ||
      el.mozMatchesSelector ||
      el.webkitMatchesSelector ||
      el.oMatchesSelector).call(el, selector)
  },

  addClass (el, className) {
    if (!el) {
      return
    }

    if (el.classList) {
      el.classList.add(className)
    } else {
      el.className += ` ${className}`
    }
  },

  removeClass (el, className) {
    if (!el) {
      return
    }

    if (el.classList) {
      el.classList.remove(className)
    } else {
      el.className = el.className.replace(new RegExp('(^|\\b)' + className.split(' ').join('|') + '(\\b|$)', 'gi'), ' ')
    }
  },

  scrollTo (el, to, duration) {
    if (duration <= 0 || !el) {
      return
    }
    const difference = to - el.scrollTop
    const perTick = difference / duration * 10
    window.setTimeout(() => {
      el.scrollTop = el.scrollTop + perTick
      if (el.scrollTop === to) {
        return
      }
      this.scrollTo(el, to, duration - 10);
    }, 10)
  }
}
