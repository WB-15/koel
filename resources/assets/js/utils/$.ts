/**
 * A utility that aims to replace jQuery for the most basic DOM methods.
 */
export const $ = {
  is: (el: Element, selector: string) => el.matches(selector),
  addClass: (el: Element | null, className: string) => el?.classList.add(className),
  removeClass: (el: Element | null, className: string) => el?.classList.remove(className),

  scrollTo (el: Element, to: number, duration: number, cb?: TAnyFunction) {
    if (duration <= 0 || !el) {
      return
    }

    const difference = to - el.scrollTop
    const perTick = difference / duration * 10

    window.setTimeout(() => {
      el.scrollTop = el.scrollTop + perTick

      if (el.scrollTop === to) {
        cb && cb()
        return
      }

      this.scrollTo(el, to, duration - 10)
    }, 10)
  }
}
