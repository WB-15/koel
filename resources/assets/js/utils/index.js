export * from './filters'
export * from './formatters'
export * from './supports'
export * from './common'
