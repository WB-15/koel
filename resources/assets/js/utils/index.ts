export * from './event'
export * from './alerts'
export * from './filters'
export * from './formatters'
export * from './supports'
export * from './common'
export * from './$'
export * from './helpers'
export * from './fileReader'
export * from './directoryReader'
