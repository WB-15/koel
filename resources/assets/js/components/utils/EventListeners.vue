<template>
  <span></span>
</template>

<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import router from '@/router'
import { authService } from '@/services'
import { playlistStore, preferenceStore, userStore } from '@/stores'
import { alerts, eventBus, forceReloadWindow } from '@/utils'

eventBus.on({
  PLAYLIST_DELETE (playlist: Playlist) {
    const destroy = async () => {
      await playlistStore.delete(playlist)
      alerts.success(`Deleted playlist "${playlist.name}."`)
      router.go('home')
    }

    if (!playlist.is_smart && !playlist.songs.length) {
      destroy()
    } else {
      alerts.confirm(`Delete the playlist "${playlist.name}"?`, destroy)
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  async LOG_OUT () {
    await userStore.logout()
    authService.destroy()
    forceReloadWindow()
  },

  'KOEL_READY': () => router.init(),

  /**
   * Hide the panel away if a main view is triggered on mobile.
   */
  'LOAD_MAIN_CONTENT': () => isMobile.phone && (preferenceStore.showExtraPanel = false)
})
</script>
