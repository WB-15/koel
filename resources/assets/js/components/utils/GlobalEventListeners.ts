/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */

import { defineComponent } from 'vue'
import { authService } from '@/services'
import { playlistFolderStore, playlistStore, userStore } from '@/stores'
import { eventBus, forceReloadWindow, requireInjection } from '@/utils'
import { RouterKey } from '@/symbols'
import { useDialogBox, useMessageToaster } from '@/composables'

export const GlobalEventListeners = defineComponent({
  setup (props, { slots }) {
    const { toastSuccess } = useMessageToaster()
    const { showConfirmDialog } = useDialogBox()
    const router = requireInjection(RouterKey)

    eventBus.on('PLAYLIST_DELETE', async playlist => {
      if (await showConfirmDialog(`Delete the playlist "${playlist.name}"?`)) {
        await playlistStore.delete(playlist)
        toastSuccess(`Playlist "${playlist.name}" deleted.`)
        router.go('home')
      }
    }).on('PLAYLIST_FOLDER_DELETE', async folder => {
      if (await showConfirmDialog(`Delete the playlist folder "${folder.name}"?`)) {
        await playlistFolderStore.delete(folder)
        toastSuccess(`Playlist folder "${folder.name}" deleted.`)
        router.go('home')
      }
    }).on('LOG_OUT', async () => {
      await userStore.logout()
      authService.destroy()
      forceReloadWindow()
    })

    return () => slots.default?.()
  }
})
