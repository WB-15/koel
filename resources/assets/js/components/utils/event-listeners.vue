<template>
  <span></span>
</template>
<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import router from '@/router'
import { auth } from '@/services'
import { playlistStore, preferenceStore, userStore } from '@/stores'
import { alerts, eventBus, forceReloadWindow } from '@/utils'

eventBus.on({
  PLAYLIST_DELETE (playlist: Playlist) {
    const destroy = async () => {
      await playlistStore.delete(playlist)
      alerts.success(`Deleted playlist "${playlist.name}."`)
      router.go('home')
    }

    if (!playlist.songs.length) {
      destroy()
    } else {
      alerts.confirm(`Delete the playlist "${playlist.name}"?`, destroy)
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  async LOG_OUT () {
    await userStore.logout()
    auth.destroy()
    forceReloadWindow()
  },

  'KOEL_READY': () => router.init(),

  /**
   * Hide the panel away if a main view is triggered on mobile.
   */
  'LOAD_MAIN_CONTENT': () => isMobile.phone && (preferenceStore.showExtraPanel = false)
})
</script>
