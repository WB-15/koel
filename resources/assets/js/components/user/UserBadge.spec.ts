import UnitTestCase from '@/__tests__/UnitTestCase'
import factory from '@/__tests__/factory'
import { expect, it } from 'vitest'
import { eventBus } from '@/utils'
import { fireEvent } from '@testing-library/vue'
import UserBadge from './UserBadge.vue'

new class extends UnitTestCase {
  private renderComponent () {
    return this.actingAs(factory<User>('user', {
      name: 'John Doe'
    })).render(UserBadge)
  }

  protected test () {
    it('renders', () => expect(this.renderComponent().html()).toMatchSnapshot())

    it('logs out', async () => {
      const emitMock = this.mock(eventBus, 'emit')
      const { getByTestId } = this.renderComponent()

      await fireEvent.click(getByTestId('btn-logout'))

      expect(emitMock).toHaveBeenCalledWith('LOG_OUT')
    })
  }
}
