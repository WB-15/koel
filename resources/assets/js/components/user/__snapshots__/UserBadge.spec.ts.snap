// Vitest Snapshot v1

exports[`renders 1`] = `<span id="userBadge" class="profile"><a class="view-profile" href="/#/profile" title="View/edit user profile"><img alt="Avatar of John Doe" src="https://gravatar.com/foo" class="avatar"><span class="name">John Doe</span></a><a class="logout control" role="button" title="Log out"><br data-testid="icon" icon="[object Object]"></a></span>`;
